animation_nodes/nodes/generic/animate_data.py

```diff
@@ -20,22 +41,81 @@ class AnimateDataNode(bpy.types.Node, AnimationNode):
 
     dataType: StringProperty(default = "Float", update = AnimationNode.refresh)
 
+    useListA: VectorizedSocket.newProperty()
+    useListB: VectorizedSocket.newProperty()
+    useListC: VectorizedSocket.newProperty()
+    useListD: VectorizedSocket.newProperty()
+
     def create(self):
-        self.newInput("Float", "Time", "time")
-        self.newInput(self.dataType, "Start", "start")
-        self.newInput(self.dataType, "End", "end")
-        self.newInput("Interpolation", "Interpolation", "interpolation", defaultDrawType = "PROPERTY_ONLY")
-        self.newInput("Float", "Duration", "duration", value = 20, minValue = 0.001)
+        self.newInput(VectorizedSocket("Float", "useListA",
+            ("Time", "time"), ("Times", "times")))
+        self.newInput(VectorizedSocket(self.dataType, ["useListB"],
+            ("Start", "start"), ("Starts", "starts")))
+        self.newInput(VectorizedSocket(self.dataType, ["useListC"],
+            ("End", "end"), ("Ends", "ends")))
+        self.newInput("Interpolation", "interpolation", defaultDrawType = "PROPERTY_ONLY")
+        self.newInput(VectorizedSocket("Float", "useListD",
+            ("Duration", "duration", dict(value = 20, minValue = 0.001)),
+            ("Durations", "durations")))
 
-        self.newOutput("Float", "Time", "outTime")
-        self.newOutput(self.dataType, "Result", "result")
+        self.newOutput(VectorizedSocket("Float", ["useListA", "useListB",
+            "useListC", "useListD"], ("Time", "outTime"), ("Times", "outTimes")))
+        self.newOutput(VectorizedSocket(self.dataType, ["useListA", "useListB", 
+            "useListC", "useListD"], ("Result", "result"), ("Results", "results")))
 
     def drawLabel(self):
         return "Animate " + self.inputs[1].dataType
 
     def getExecutionCode(self, required):
-        yield "finalDuration = max(duration, 0.0001)"
-        yield "influence = max(min(time / finalDuration, 1.0), 0.0)"
-        yield "influence = interpolation(influence)"
-        yield getMixCode(self.dataType, "start", "end", "influence", "result")
-        yield "outTime = time - finalDuration"
+        if any([self.useListA, self.useListB, self.useListC, self.useListD]):
+            args = ", ".join(socket.identifier for socket in self.inputs)
+            yield "results, outTimes = self.executeDataList({})".format(args)
+
+        else:
+            yield "finalDuration = max(duration, 0.0001)"
+            yield "influence = max(min(time / finalDuration, 1.0), 0.0)"
+            yield "influence = interpolation(influence)"
+            yield getMixCode(self.dataType, "start", "end", "influence", "result")
+            yield "outTime = time - finalDuration"
+
+    def executeDataList(self, time, start, end, interpolation, duration):
```

Comment:
You can probably just add that to the execution code directly. This way, you can dynamically select the virtual list type and the needed execution function directly.

Reply:
sorry, Can U make more detailed about this? how to wirite this ? Thank you!

Reply:
maybe i guess you meant like this:
 ```
def getExecutionCode(self, required):
        if any([self.useListA, self.useListB, self.useListC, self.useListD]):
            args = ", ".join(socket.identifier for socket in self.inputs)
            yield f"time, start, end, interpolation, duration = {args}"
            yield f"starts, ends = AN.data_structures.Virtual{self.dataType}List.createMultiple((start, 0), (end, 0))"
            yield "amount = AN.data_structures.VirtualDoubleList.getMaxRealLength(times, starts, ends, durations)"
            yield "factors = AN.nodes.generic.c_utils.calculateInfluenceList(times, interpolation, durations, amount)"
            yield "_factors = AN.data_structures.VirtualDoubleList.create(factors, 0)"
            yield f"outTimes = AN.nodes.generic.c_utils.executeTimeList(times, durations, amount)"
            yield "results = AN.nodes.generic.c_utils.mix{self.dataType}Lists(starts, ends, _factors, amount)"
```

Reply:
Yes. Something like that. Though this probably will not work because `Virtual{self.dataType}List` will not be correct for vector lists and maybe others. So, a dictionary mapping the data type to the appropriate virtual list can be created outside the node class and used to retrieve the name of the virtual list.

Also, to avoid this line:
```
args = ", ".join(socket.identifier for socket in self.inputs)
```
You can just use the same identifier for both base and list inputs. So instead of `("Time", "time"), ("Times", "times")`, you do `("Time", "time"), ("Times", "time")`. Notice the name stays the same, but the identifier is made identical.

Reply:
sorry, i don't know how to use a dictionary mapping the data type, can U give me example? Thank you!

Reply:
@kdsin23 Outside the class, you should define the following dictionary:

```python
dataTypeToVirtualListMapping = {
    "Float" : "VirtualDoubleList",
    "Vector" : "VirtualVector3DList",
    ...
}
```

Then, in the execution code you do:

```
f"AN.data_structures.{dataTypeToVirtualListMapping[self.dataType]}.createMultiple((start, 0), (end, 0))"
```